Hit Loading Speedup

GNiendorf opened the pull request:
This PR speeds up hit loading by moving calculations done in addHitToEvent to the GPU. I decided to not make changes to the hit preloading code until we decide if we want to keep it in the codebase. I did however move the timing/validation scripts back over to not using the preloading flag by default.

Validation plots:
https://www.classe.cornell.edu/~gsn27/www/PR181/

Timing with no preloading:
![PRTiming](https://user-images.githubusercontent.com/25272611/176475642-9915ceae-fccf-4847-8238-b977962a6145.png)

Also, the "Real Time" for 200 events drops from 42s (w/ preloading) to 30s (w/o preloading) under this PR for single stream.

pwittich commented:
When @tresreid comes back we can decide if this should be ported to the hit loading or if we should get rid of it.

tresreid commented:
I think it makes sense to drop the hit preloading. I think this is a better version overall. Given the timing results, the preloading isn't necessary anymore

GNiendorf commented:
Validation plots
https://www.classe.cornell.edu/~gsn27/www/PR181_4/

Timing
![removepre](https://user-images.githubusercontent.com/25272611/176955469-aca8d96e-2a90-4497-9147-95505d897dba.png)

I will upload comparison validation plots to the current master soon.

GNiendorf commented:
Comparison plots:
https://www.classe.cornell.edu/~gsn27/www/PR181_comp/

![TC_AllTypes__ptzoom](https://user-images.githubusercontent.com/25272611/176971551-5391c846-a3e8-4a70-9ef7-9bfcf461bfd2.png)
![TC_AllTypes__pt](https://user-images.githubusercontent.com/25272611/176971552-d022599b-6db3-4d88-85a5-a0f157160551.png)

GNiendorf commented:
I checked all of my variables on the GPU against their CPU-calculated values and couldn't find any major outliers. I think what's going on is that when I calculate a certain variable on the GPU, like sqrt(hit_x*hit_x+hit_y*hit_y), it differs slightly from its CPU version because CUDA's math functions are calculated differently than the standard library's math functions. I also noticed that one variable which is calculated on both the CPU and GPU with the same _global_ function agrees much better than the other variables (up to the last decimal compared to only 4-5 decimal places). To check that this is causing the efficiency differences, I made a hybrid CPU/GPU version of the code where I first calculate the variables on the CPU and upload them to the GPU like what is done on the master branch, but then override them with my GPU kernels. If I comment out only the variables that differ by math precision (but override everything else), then I get much better agreement to master.

https://www.classe.cornell.edu/~gsn27/www/PR181_comp2/

![TC_AllTypes__ptzoom2](https://user-images.githubusercontent.com/25272611/176980123-820e5d1b-6ab9-4c27-a696-d1a710070d7c.png)

slava77 commented:
> Comparison plots: https://www.classe.cornell.edu/~gsn27/www/PR181_comp/
> 
> The differences are small, but I will try to see where they are coming from.
> 
> ![TC_AllTypes__ptzoom](https://user-images.githubusercontent.com/25272611/176971551-5391c846-a3e8-4a70-9ef7-9bfcf461bfd2.png)

Does this comparison use all of the events available? If the differences are from a random shuffle, it would be best to compare with the smallest uncertainty (largest sample).

slava77 commented on SDL/Event.cu:
```diff
@@ -574,464 +566,151 @@ void SDL::Event::resetObjectsInModule()
-
+    // No stream synchronize needed after second kernel call. Saves ~100 us.
+    moduleRangesKernel<<<MAX_BLOCKS,256,0,stream>>>(nLowerModules, modulesInGPU, hitsInGPU);
+    //std::cout << cudaGetLastError() << std::endl;
```
Is the sync not needed because the next operation is `addPixelSegmentToEvent`), which does not rely on `hitsInGPU.hitRanges*`  (the only values modified in this kernel)?

It would be better to be clear about this, to reduce a chance of errors in the future

GNiendorf commented:
I removed the extra semicolon and explained why a sync is not needed after the second kernel. I also moved hitsInGPU to the host caching allocator, since previously the hit preloading code was causing an error that prevented it from being moved.

Validation plots with all events available:
https://www.classe.cornell.edu/~gsn27/www/PR181_postclean1/

Current master with all events available:
https://www.classe.cornell.edu/~gsn27/www/PR181_masterfull/

Comparison with all events available:
https://www.classe.cornell.edu/~gsn27/www/PR181_compfull/

New Timing:
![t1](https://user-images.githubusercontent.com/25272611/177054116-bcac463f-8b64-4c2e-9e7c-432b595744c4.png)

![zoom1](https://user-images.githubusercontent.com/25272611/177056730-1f0951da-f863-4226-ae2c-8166fc1310a7.png)

GNiendorf commented:
Peter pointed out that these should be cosf and sinf since the inputs are floats. I tried sincosf but that gave random runtime errors. Doesn't affect the timing substantially, but it's still good to move them over to the correct functions.

pwittich commented on SDL/Event.cu:
```diff
@@ -574,464 +564,154 @@ void SDL::Event::resetObjectsInModule()
+            hitsInGPU->lowEdgeXs[ihit] = ihit_x - 2.5 * cos_phi;
+            float sin_phi = sinf(phi);
+            hitsInGPU->highEdgeYs[ihit] = ihit_y + 2.5 * sin_phi;
+            hitsInGPU->lowEdgeYs[ihit] = ihit_y - 2.5 * sin_phi;
```
All these float should be put in `2.5f` to avoid the conversion to doubles, otherwise switching to `cosf` and `sinf` probably doesn’t matter.

slava77 commented:
> Hey @slava77 can you check if my changes address all the concerns you had when you get a chance? Thank you!

yes, my comments were resolved. Thanks.
